pkg/apis/core/validation/validation.go

```diff
@@ -7253,3 +7289,88 @@ func ValidatePodAffinityTermSelector(podAffinityTerm core.PodAffinityTerm, allow
 	allErrs = append(allErrs, unversionedvalidation.ValidateLabelSelector(podAffinityTerm.NamespaceSelector, labelSelectorValidationOptions, fldPath.Child("namespaceSelector"))...)
 	return allErrs
 }
+
+func validateNodeSelectorMutation(fldPath *field.Path, oldNodeSelector, newNodeSelector map[string]string) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node selector was nil, anything can be set.
+	if oldNodeSelector == nil {
+		return allErrs
+	}
+	// Validate no existing node selectors were deleted or mutated.
+	for k, v1 := range oldNodeSelector {
+		if v2, ok := newNodeSelector[k]; !ok || v1 != v2 {
+			allErrs = append(allErrs, field.Invalid(fldPath, newNodeSelector, "only additions to spec.nodeSelector are allowed (no mutations or deletions)"))
+			return allErrs
+		}
+	}
+	return allErrs
+}
+
+func validateNodeAffinityMutation(nodeAffinityPath *field.Path, oldNodeAffinity, newNodeAffinity *core.NodeAffinity) field.ErrorList {
+	var allErrs field.ErrorList
+	// If old node affinity was nil, anything can be set.
+	if oldNodeAffinity == nil {
+		return allErrs
+	}
+
+	// RequiredDuringSchedulingIgnoredDuringExecution must not be nil for both old and new nodeAffinity
+	// in order to perform this validation.
+	if oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil || newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution == nil {
+		return allErrs
+	}
+
+	// NodeSelectorTerms are semantically OR-ed, so adding a new one would
+	// broaden the constraint rather than limit it further.
+	if len(oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) != len(newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms) {
+		return append(allErrs, field.Invalid(nodeAffinityPath.Child("requiredDuringSchedulingIgnoredDuringExecution").Child("nodeSelectorTerms"), newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms, "no additions to non-empty NodeSelectorTerms list are allowed"))
+	}
+
+	// For requiredDuringSchedulingIgnoredDuringExecution, if old NodeSelectorTerms
+	// was empty, anything can be set. If non-empty, only additions of NodeSelectorRequirements
+	// to matchExpressions or fieldExpressions are allowed.
+	for _, oldTerm := range oldNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
+		found := false
+		for _, newTerm := range newNodeAffinity.RequiredDuringSchedulingIgnoredDuringExecution.NodeSelectorTerms {
```

Comment:
This lets oldTerms change locations in the list... is there a reason for that flexibility?

I expected something more like:
```go
for i := range oldTerms {
  if !validateNodeSelectorTermHasOnlyAdditions(newTerms[i], oldTerms[i]) {
     ... error
  }
}
```

Reply:
yes, order of lists is deterministic (previously we required DeepEqual to return true on these since they were completely immutable)

Reply:
Sounds good to me too; I also thought about suggesting this to simplify the code, but the KEP didn't discuss that the order should continue to be deterministic...